src/api/web3.ts

```diff
@@ -0,0 +1,210 @@
+import Web3 from 'web3'
+import { Block } from 'web3-eth'
+import { provider } from 'web3-providers'
+import { TransactionReceipt, Transaction } from 'web3-core'
+
+import { windowLoaded } from './utils'
+
+declare global {
+    interface Window {
+        ethereum: provider & {
+            enable: () => Promise<string[]>
+            on?: () => EventListener
+        }
+        web3: Web3
+    }
+}
+
+export interface Web3FrontendAPI {
+    web3: Web3
+    getCurrentAccount: () => Promise<string>
+    getAccounts: () => Promise<string[]>
+    getBalance: (account?: string) => Promise<string>
+    getBlock: (blockNumberOrString: string | number) => Promise<Partial<Block>>
+    getNetwork: () => Promise<string | number>
+    getNetworkName: (id?: number | string) => Promise<string>
+    getNetworkProviderURL: (id: number | string) => string
+    getTransaction: (txHash: string) => Promise<Transaction>
+    getTransactionReceipt: (txHash: string) => Promise<TransactionReceipt>
+    // Utils
+    isAddress: (address: string) => boolean
+    isReadOnly: () => boolean
+    // Provider Specific
+    resetProvider(): void
+    setProvider(provider: string): void
+}
+
+export interface Web3SetupOptions {
+    customProvider?: provider | string
+    reloadOnAccountsChange?: boolean
+}
+
+let web3API: Web3FrontendAPI
+
+let readOnly: boolean
+
+export const getWeb3API = async (options: Web3SetupOptions = {}, force?: boolean): Promise<Web3FrontendAPI> => {
+    if (!force && web3API) return web3API
+
+    return setupWeb3(options)
+}
+
+const getProvider = async ({ customProvider, reloadOnAccountsChange }: Web3SetupOptions): Promise<Web3> => {
+    let provider: Web3
+
+    // wait window load then check window globals...
+    await windowLoaded
+
+    // Use Custom Provider if passed in
+    if (customProvider) provider = new Web3(customProvider)
+    // No Custom Provider? Check window.ethereum
+    else if (window && window.ethereum) {
+        provider = new Web3(window.ethereum)
+
+        // enable Provider, if possible
+        await window.ethereum.enable().catch((error: Error) => {
+            throw new Error(error.message)
+        })
+
+        if (window.ethereum.on && reloadOnAccountsChange) {
```

Comment:
I think, it's better API for this function to change `reloadOnAccountsChange` for `onAccountsChange`.

This way, we can implement any update logic that react to account changes. Of course the trivial one too: 
```js
getProvider({
   customProvider,
   onAccountsChange: () => window.location.reload()`
})
```

Reply:
I don't understand, are you saying make this default to yes?

Reply:
Actually, agree. Leave it to the user to act on known events
Even better would be
```
onAccountsChange: (provider, payload) => window.location.reload()
```
Where payload is actual data for that event

Reply:
Why would you expose the provider also in the callback?
It's not better just:
`onAccountsChange: accounts => window.location.reload()`